pkg/controller/networkpolicy/controller_querier_replier.go

```diff
@@ -0,0 +1,162 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package networkpolicy provides NetworkPolicyController implementation to manage
+// and synchronize the Pods and Namespaces affected by Network Policies and enforce
+// their rules.
+
+package networkpolicy
+
+import (
+	networkingv1beta1 "github.com/vmware-tanzu/antrea/pkg/apis/networking/v1beta1"
+	"github.com/vmware-tanzu/antrea/pkg/apiserver/storage"
+	v1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+
+	antreatypes "github.com/vmware-tanzu/antrea/pkg/controller/types"
+)
+
+type EndpointQuerier interface {
+	QueryNetworkPolicies(namespace string, podName string) *EndpointQueryResponse
+}
+
+// EndpointQueryReplier is responsible for handling query requests from antctl query
+type EndpointQueryReplier struct {
+	// addressGroupStore is the storage where the populated Address Groups are stored.
+	addressGroupStore storage.Interface
+	// appliedToGroupStore is the storage where the populated AppliedTo Groups are stored.
+	appliedToGroupStore storage.Interface
+	// internalNetworkPolicyStore is the storage where the populated internal Network Policy are stored.
+	internalNetworkPolicyStore storage.Interface
+	// podInformer is used to check existence of a selected pod
+	podInformer coreinformers.PodInformer
+}
+
+// EndpointQueryResponse is the reply struct for QueryNetworkPolicies
+type EndpointQueryResponse struct {
+	Endpoints []Endpoint
+	Error error
+}
+
+// Endpoint holds response information for an endpoint following a query
+type Endpoint struct {
+	Namespace string
+	Name string
+	Policies []Policy
+	Rules []Rule
+}
+
+type PolicyRef struct {
+	Namespace string
+	Name string
+	UID types.UID
+}
+
+// Policy holds network policy information to be relayed to client following query endpoint
+type Policy struct {
+	PolicyRef
+	selector metav1.LabelSelector
+}
+
+// Rule holds
+type Rule struct {
+	PolicyRef
+	Direction networkingv1beta1.Direction
+	RuleIndex int
+	Ports []v1.NetworkPolicyPort
+}
+
+// NewNetworkPolicyController returns a new *NetworkPolicyController.
+func NewEndpointQueryReplier(
+	addressGroupStore storage.Interface,
+	appliedToGroupStore storage.Interface,
+	internalNetworkPolicyStore storage.Interface,
+	podInformer coreinformers.PodInformer) *EndpointQueryReplier {
+	n := &EndpointQueryReplier{
+		addressGroupStore:          addressGroupStore,
+		appliedToGroupStore:        appliedToGroupStore,
+		internalNetworkPolicyStore: internalNetworkPolicyStore,
+		podInformer: podInformer,
+	}
+	return n
+}
+
+//Query functions
+func (eq EndpointQueryReplier) QueryNetworkPolicies(namespace string, podName string) *EndpointQueryResponse {
```

Comment:
move the error out of `EndpointQueryResponse` and return `(*EndpointQueryResponse, error)`

Reply:
BTW in the handler you can tell that an endpoint is "not found" if `Endpoints` is `nil` / empty. I wouldn't return an error in this case (in this function), just a `nil` list.

Reply:
You can inspect the k8s error returned by `Get` and determine if it's `NotFound` or something else